### ISSUE TITLE
DEP: Hybrid Worker — Idle Prefill GPU Utilization for Overflow Decode

### ISSUE THREAD

**Issue by mcpark84** [ai-dynamo/dynamo#9474]
### Area

backend-vllm

### Summary

  Introduce a **Hybrid Worker** mode (`--disaggregation-mode hybrid`) that allows a single vLLM instance to act simultaneously as a prefill worker and an overflow decode target. In disaggregated deployments, prefill workers are typically underutilized (20–40% GPU utilization) due to bursty prefill patterns. This DEP proposes reusing that idle capacity to absorb overflow decode requests when dedicated decode workers are saturated, improving overall system throughput without adding hardware.                                                                                                                                                                                                                                              
                                                                             
  To route overflow decode traffic selectively, this DEP introduces a **2-tier decode routing model**. Dedicated decode workers form **tier-0** — the primary pool that handles all decode traffic under normal load. Hybrid workers register as **tier-1** — a secondary pool that the KV router activates only when every tier-0 worker has exceeded a configurable KV cache fill threshold (`DYN_HYBRID_KV_FILL_RATIO`). Under normal conditions, hybrid workers remain exclusively focused on prefill; tier-1 exists solely to absorb the overflow that would otherwise be rejected or queued. This tiered design ensures decode workers are always preferred, while the hybrid's idle GPU capacity acts as a transparent elastic buffer. 

### Motivation

  In a standard disaggregated deployment, prefill and decode workers are statically assigned to separate GPU instances. This creates two compounding inefficiencies:                                                                                     
   
  1. **Prefill GPU underutilization**: Prefill computation is bursty — heavy during request spikes and nearly idle between them. GPU utilization on prefill workers typically sits at 20–40%, leaving the majority of compute capacity unused.           
                                                                             
  2. **Decode throughput bottleneck**: Under sustained high concurrency, dedicated decode workers become the system throughput limit. There is no mechanism to route overflow decode traffic elsewhere, so requests either queue or fail.                
                                                                             
  These two problems are complementary: the resource that is underused (prefill GPU) is exactly what is needed to solve the resource that is scarce (decode capacity). A hybrid worker that can switch roles between prefill and overflow decode directly addresses both inefficiencies without requiring additional hardware.   

### Proposal

### Two-Endpoint Architecture

  Each hybrid worker pod registers two endpoints sharing the same vLLM engine instance:

  | Endpoint | NAME | ModelType | decode_tier | Handler |
  |---|---|---|---|---|
  | `prefill_endpoint` | `generate_prefill` | Prefill | — | `handler.generate()` |
  | `decode_overflow_endpoint` | `generate` | Chat\|Completions | 1 | `handler.overflow_generate()` |

  `decode_overflow_endpoint` uses the same endpoint NAME (`generate`) and component (`backend`) as dedicated decode workers, placing it in the same `workers_with_configs` pool. This is the mechanism by which the KV router can include hybrid workers in routing decisions without changes to the upstream discovery pipeline.

  The two endpoints use distinct NAMEs to avoid model card key collisions: discovery keys are scoped as `{namespace}/{component}/{endpoint}/{instance_id}`, so registering two model cards (Prefill + Chat|Completions) under the same NAME from the same instance would result in the second overwriting the first.

  ### 2-Tier Decode Routing                                                  

  `scheduler.rs` is extended with a 2-tier routing policy. Every routing decision first classifies workers by the presence of `decode_tier == 1` in their model card `runtime_data`:

  - **Tier-0**: workers without `decode_tier` or with `decode_tier=0` — always the first choice
  - **Tier-1**: workers with `decode_tier=1` (hybrid) — included only when all tier-0 workers exceed the fill threshold
                                                                                                                                                                                                                                                         
  The tier-1 activation check evaluates whether every tier-0 worker satisfies:
                                                                                                                                                                                                                                                         
  active_blocks >= total_kv_blocks × DYN_HYBRID_KV_FILL_RATIO                                                                                                                                                                                            
   
  When no hybrid workers are present, the entire 2-tier code path is skipped, preserving the existing routing behavior exactly.                                                                                                                          
                                                                             
  ### Cost Function Extensions

  Two penalties are added to hybrid worker logits to prevent overloading:                                                                                                                                                                                
   
  **Capacity penalty** (`DYN_HYBRID_COST_PENALTY`): Adds `total_kv_blocks × fill_ratio` to the hybrid logit. This ensures an idle hybrid (with near-zero tracked decode blocks) is never artificially cheaper than a moderately loaded decode worker. The same `fill_ratio` value governs both the tier-1 activation threshold and the cost-equivalence point, so a single knob controls both behaviors.
                                                                                                                                                                                                                                                         
  **Prefill load penalty** (automatic): `PrefillGuard` RAII in `prefill_router.rs` accumulates the actual ISL token count for all in-flight prefill requests per worker in a process-global `DashMap<worker_id, AtomicI64>`. The decode scheduler reads this value and adds `active_prefill_tokens / block_size` to the hybrid logit, penalizing hybrids that are actively prefilling without requiring any fixed-ISL configuration.
                                                                                                                                                                                                                                                         
  ### Self-Source Filter                                                     

  When the same hybrid instance both prefills a request and receives it as overflow decode, NIXL self-transfer fails. This is prevented at two levels:                                                                                                   
   
  1. **Router level** (`DYN_HYBRID_SELF_ROUTING=no`, default): `prefill_router.rs` peeks the selected prefill worker ID before execution (`update_states=false`) and injects it into `RouterConfigOverride.prefill_worker_id`. The decode scheduler then excludes that worker ID from tier-1 candidates for that request.           
                                                                                                                                                                                                                                                         
  2. **Handler level** (safety fallback): `overflow_generate()` checks whether `source_instance_id` in `disaggregated_params` matches its own instance ID. If so, `kv_transfer_params` is stripped, causing vLLM to fall back to aggregated mode with a prefix cache hit at near-zero cost.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

### Alternate Solutions

_No response_

### Requirements

_No response_

### References

<img width="889" height="486" alt="Image" src="https://github.com/user-attachments/assets/472ad7bd-9d5d-43b7-9938-5e52a21e73b8" />

The graph shows throughput (token throughput/sec) vs. e2e latency. The blue line represents the hybrid configuration (2 hybrid workers + 2 decode workers, Qwen3-30B, ISL=4k, OSL=4k, KV fill ratio=0.1), while the red line serves as the reference baseline (2 prefill workers + 2 decode workers under identical conditions). The hybrid configuration demonstrates a clear throughput advantage over the baseline, with the gap widening as concurrency increases.

**Comment by athreesh** [comment]
@yunzhoul-nv any status on who is best to take a look at this?

**Comment by jthomson04** [comment]
@mcpark84 A couple comments/questions:

1. The disagg benchmark you share doesn't seem to be properly rate matched. 4k/4k ISL/OSL with 2p2d seems like it excessively skews towards prefill. This would make the hybrid approach look artificially good. Could you run a baseline with 4 agg workers to compare these curves against?
2. The kv fill ratio being set in the benchmarks is 0.1, so prefill is practically running as an agg worker at higher concurrency. That doesn't align with the stated goal of only going to the overflow prefill workers when decode is saturated. 10% kv utilization doesn't indicate saturation.
3. This seems like a good application for planner. SLA-aware autoscaling could address the scenario of prefill being mostly idle while decode is saturated.

**Comment by mcpark84** [comment]
@jthomson04 Thanks for your response.
For the question 1 regarding Rate matching / 4-agg baseline, We will include 4-agg baseline results in the next update. But I'm sure 4-agg result would be better. That said, we'd like to clarify the goal of this DEP: it is not to claim throughput superiority over 4-agg, but rather to address a specific operational scenario — one where a disaggregated deployment is already in place, and a temporary mismatch arises where prefill nodes are idle while decode nodes are congested.
 In our experience serving 30–80B models, we have rarely observed disaggregated PD-separation outperforming aggregated serving in raw throughput. The primary advantage of disaggregation, in our view, is enabling more predictable TTFT and ITL SLA compliance under high concurrency — for example, through tools like Planner.
 Given that context, the Hybrid Worker is best understood as one option for absorbing decode burst using already-deployed, idle prefill GPUs — without provisioning additional hardware. The intended comparison baseline is not 4-agg, but rather the same hardware in a pure disagg setup.

For the question 2 regarding kv ratio 0.1, You are correct. A fill ratio of 0.1 does not represent a congested decode state — at that threshold, overflow routing activates well before decode workers are meaningfully loaded, causing the hybrid worker to behave more like a  standard aggregated worker at higher concurrency.
 The 0.1 setting was used as an illustrative example to confirm that the overflow routing mechanism was functioning correctly, not as a representation of the intended operating point. The designed default is 0.8, where overflow to hybrid workers triggers only when all decode workers have exceeded 80% KV utilization. With fill_ratio=0.8, performance is identical to the disagg baseline at low concurrency, and the hybrid routing advantage is only observed under high concurrency when decode workers are genuinely saturated — which aligns with the intended design goal.

**Comment by karen-sy** [comment]
Hi @mcpark84 , thank you for the interesting proposal, and sorry for the delayed feedback.

1. Baselines: I'd also like to see the 4-agg baseline, as well as 1p3d disagg. I wonder how much gain this hybrid routing approach buys us that can't be achieved with p/d ratematching, e.g. via Planner. To better understand your effective prefill workload, could you also share stats on your kv cache hit %? If effective ISL is very low there might not be a strong reason to use disagg over agg.
2. Disagg vs Agg: I agree that disagg generally "enables more predictable SLA compliance under high concurrency." Could you also share the same latency vs tput plots (with agg vs disagg vs hybrid disagg) for fill_ratio=0.8, at a couple high-concurrency scenarios you've tested? I'm curious how much the lift the approach provides in that case. 

As an aside, we are actively working on "conditional disagg" ([branch](https://github.com/ai-dynamo/dynamo/tree/karenc/conditional-prefill-v1)) which is actually aimed at an inverse problem, i.e. prefill is too loaded, decode too idle. Idea is similar to yours -- we make the decode pool hybrid, such that when prefill workers are too busy decode workers can absorb its workload.

**Comment by mcpark84** [comment]
Hi @karen-sy. Thanks for your comments. I would do extra experiments for your request and share you soon. In terms of "conditional disagg", I know it's on roadmap. I think "conditional disagg" and hybrid worker can be working at complementary manner.

**Comment by mcpark84** [comment]
Hi @karen-sy Please find the updated graphs attached, which now include 4-agg and 1p3d disagg as baselines alongside the 1h3d (1 hybrid + 3 decode) results at fill_ratio=0.2 and 0.4.

A note on the fill_ratio selection: in our current experimental setup, fill_ratio=0.8 does not trigger hybrid overflow routing, as decode KV utilization does not reach 80% within the tested concurrency range. We therefore used fill_ratio=0.2 and 0.4 as operating points to observe hybrid behavior, which provides a clearer view of the routing boundary. As shown in the graphs, fill_ratio=0.4 begins to influence routing only at concurrency=384, while fill_ratio=0.2 starts affecting results from concurrency=256.

In terms of overall throughput ranking across the tested configurations: 4-agg performs best, 1p3d disagg performs worst, and 1h3d sits between the two — with a measurable performance gain over 1p3d at high concurrency, depending on fill_ratio. This is consistent with our stated goal: the hybrid approach is not intended to outperform aggregated serving, but rather to recover throughput compared to a pure disaggregated setup when prefill workers would otherwise be idle.

The ITL graph is also included. ITL differences across configurations are minimal.


<img width="974" height="536" alt="Image" src="https://github.com/user-attachments/assets/040129bc-b000-4346-ab5c-78f16712d7ee" />

<img width="974" height="535" alt="Image" src="https://github.com/user-attachments/assets/e6886148-26a3-4e50-b0eb-59aae4a7df92" />

**Comment by mcpark84** [comment]
Also regarding the KV cache hit rate: the measured values align closely with the intended 50% hit rate in our simulation dataset. We tracked this using the following Prometheus query:
rate(vllm:prefix_cache_hits_total[5m]) / rate(vllm:prefix_cache_queries_total[5m])


<img width="557" height="334" alt="Image" src="https://github.com/user-attachments/assets/33ce6ab1-5f55-4e19-b7f7-561961669ef5" />

The observed hit rate was approximately 50%, matching the simulation's target. This also means that in self-source overflow scenarios (where the hybrid worker both prefills and decodes the same request in aggregated mode), the prefix cache effectively eliminates redundant recomputation, which is consistent with our design assumption.

**Comment by karen-sy** [comment]
Nice, thanks for the experiments! Nvidia is on companywide break this week, so I'll take a closer look / give it more thought early next week. As you said, there should be a nice way to integrate this idea into conditional disagg (in a way that makes it birectional).
